### PR TITLE
do not call profile code from C main()

### PR DESCRIPTION
**Pull request by WalterBright**
The profiler requires TLS to be working properly, and it is not when OSX's C main() gets control. So don't call the profiler then.

Required for:

https://github.com/D-Programming-Language/druntime/pull/1233

### CODE REVIEW

**Comment by MartinNowak** [comment]
I think we should rather check in druntime, whether or not TLS is already initialized, and do an early exit otherwise.
- Then you can also properly use this on OSX with a C main and rt_init.
- There will be more platforms with emulated TLS.
- Silently disabling this on some intransparant condition is highly confusing.

**Comment by WalterBright** [comment]
I find the TLS code in druntime to be extremely hard to follow. If you can recommend how to do this, I'd appreciate it.

**Comment by WalterBright** [comment]
Actually, the way the profiler works, it would be significantly hampered by putting in a check for TLS in every call to the profiler. I don't believe it is a big problem to simply not profile C main() in OSX, nor do I believe it is highly confusing because 1) it is documented and 2) the C main() is not a user entry point, it is specifically used to construct the D runtime. Profiling it doesn't even make much sense.

File: src/func.c [hunk 1]
```diff
@@ -3549,6 +3549,12 @@ bool FuncDeclaration::isMain()
         linkage != LINKc && !isMember() && !isNested();
 }
 
+bool FuncDeclaration::isCMain()
+{
+    return ident == Id::main &&
+        linkage == LINKc && !isMember() && !isNested();
```
Comment (andralex) [inline src/func.c]: Should this have a check that it's not static?

File: src/glue.c [hunk 2]
```diff
@@ -1174,8 +1174,12 @@ void FuncDeclaration_toObjFile(FuncDeclaration *fd, bool multiobj)
          * 2. impact on function inlining
          * 3. what to do when writing out .di files, or other pretty printing
          */
-        if (global.params.trace)
+        if (global.params.trace &&
+            !(global.params.isOSX && fd->isCMain()))
```
Comment (andralex) [inline src/glue.c]: Do we ever want to trace C main? Probably not, so I guess the isOSX test may be eliminated.

File: src/func.c (hunk 1 above)
Comment (WalterBright) [inline src/func.c]: The !isMember() takes care of that.

File: src/glue.c (hunk 2 above)
Comment (WalterBright) [inline src/glue.c]: It's neither here nor there. I'll remove the check.

File: src/func.c (hunk 1 above)
Comment (andralex) [inline src/func.c]: k thx

**Comment by andralex** [comment]
lgtm, will toggle merge but you need to fix failures first - thx

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
Sure, there is little reason to profile main and yes not profiling C main fixes the immediate problems on OSX, but doing an unrelated change instead of fixing the actual issue is a Hack and adds more technical debt.
For example it works now to profile a normal app on OSX, but if you create a D library with a C interface you have a different entry point, and profiling won't work.

> Actually, the way the profiler works, it would be significantly hampered by putting in a check for TLS in every call to the profiler.

What's the problem with checking a boolean?

**Comment by MartinNowak** [comment]
> If you can recommend how to do this, I'd appreciate it.

We already have a boolean flag called [_isRuntimeInitialized](https://github.com/D-Programming-Language/druntime/blob/61ba4b8d3c0052065c17ffc8eef4f11496f3db3e/src/rt/sections_osx.d#L68) that you can use.

**Comment by MartinNowak** [comment]
See https://github.com/D-Programming-Language/druntime/pull/1245.
